src/Halogen/VDom/DOM/Prop.purs

```diff
@@ -0,0 +1,238 @@
+  map f (Created a) = Created (f a)
+  map f (Removed a) = Removed (f a)
+
+foreign import data PropValue ∷ *
+
+propFromString ∷ String → PropValue
+propFromString = unsafeCoerce
+
+propFromBoolean ∷ Boolean → PropValue
+propFromBoolean = unsafeCoerce
+
+propFromInt ∷ Int → PropValue
+propFromInt = unsafeCoerce
+
+propFromNumber ∷ Number → PropValue
+propFromNumber = unsafeCoerce
+
+type PropEff eff a =
+  Eff (dom ∷ DOM, ref ∷ REF | eff) a
+
+buildProp
+  ∷ ∀ eff a
+  . (a → Eff (ref ∷ REF, dom ∷ DOM | eff) Unit)
+  → DOM.Element
+  → V.VDomMachine (ref ∷ REF, dom ∷ DOM | eff) (Array (Prop a)) Unit
+buildProp emit el = render
+  where
+  render ps1 = do
+    events ← newMutMap
+    ps1' ← Fn.runFn3 strMapWithIxE ps1 propToStrKey (applyProp events)
+    pure
+      (V.Step unit
+        (Fn.runFn2 patch (unsafeFreeze events) ps1')
+        (done ps1'))
+
+  patch = Fn.mkFn2 \prevEvents ps1 → \ps2 → do
+    events ← newMutMap
+    let
+      onThese = Fn.runFn2 diffProp prevEvents events
+      onThis = removeProp prevEvents
+      onThat = applyProp events
+    ps2' ← Fn.runFn6 diffWithKeyAndIxE ps1 ps2 propToStrKey onThese onThis onThat
+    pure
+      (V.Step unit
+        (Fn.runFn2 patch (unsafeFreeze events) ps2')
+        (done ps2'))
+
+  done ps = do
+    case StrMap.lookup "ref" ps of
+      Just (Ref f) → do
+        mbEmit (f (Removed (unsafeElementToHTMLElement el)))
+      _ → do
+        effUnit
+
+  mbEmit =
+    maybe effUnit emit
+
+  applyProp events = Fn.mkFn3 \_ _ v →
+    case v of
+      Attribute ns attr val → do
+        Fn.runFn4 setAttribute (toNullable ns) attr val el
+        pure v
+      Property prop val → do
+        Fn.runFn3 setProperty prop val el
+        pure v
+      Handler (DOM.EventType ty) f → do
+        ref ← Ref.newRef f
+        let
+          listener = DOM.eventListener \ev → do
+            f' ← Ref.readRef ref
+            mbEmit (f' ev)
+        Fn.runFn3 pokeMutMap ty (Tuple listener ref) events
+        Fn.runFn3 addEventListener ty listener el
+        pure v
+      Ref f → do
+        mbEmit (f (Created (unsafeElementToHTMLElement el)))
+        pure v
+
+  diffProp = Fn.mkFn2 \prevEvents events → Fn.mkFn4 \_ _ v1 v2 →
+    case v1, v2 of
+      Attribute _ _ val1, Attribute ns2 attr2 val2 →
+        case val1 /= val2 of
+          true → do
+            Fn.runFn4 setAttribute (toNullable ns2) attr2 val2 el
+            pure v2
+          _ →
+            effPure v2
+      Property _ val1, Property prop2 val2 →
+        case Fn.runFn2 refEq val1 val2, prop2 == "value" of
+          true, _ →
+            effPure v2
+          _, false → do
+            Fn.runFn3 setProperty prop2 val2 el
+            pure v2
+          _, _ → do
+            elVal ← Fn.runFn2 getProperty "value" el
+            case not (Fn.runFn2 refEq elVal val2) of
+              true → do
+                Fn.runFn3 setProperty prop2 val2 el
+                pure v2
+              _ →
+                pure v2
```

Comment:
Some of these cases are to keep PS from generating wacky code with if/then/elses.